mypy/semanal.py

```diff
@@ -4295,7 +4295,8 @@ def add_symbol_table_node(self,
                 if not (isinstance(new, (FuncDef, Decorator))
                         and self.set_original_def(old, new)):
                     self.name_already_defined(name, context, existing)
-        elif name not in self.missing_names and '*' not in self.missing_names:
+        elif (not (name in self.missing_names and symbol.kind != MDEF)
```

Comment:
Hmm I'm not sure if this change is safe. Why are you skipping the check for `MDEF`?

Reply:
In https://github.com/python/mypy/pull/6257 you ensure that symbol table additions are skipped if they are defining something that was already defined, but the previous definition couldn't be resolved. 

You mentioned that you don't track the scopes for simplicity, and that extra deferrals should occur in order to correct this. I'm assuming in this case that these deferrals aren't happening as we'd expect.

My reasoning here was that because in the case of Typeddicts (and NamedTuples too?) can only have members that collisions wouldn't be able to take place in the scope of the class. It seems like we either want to treat this as a special case and only have this condition for these types of class or investigate further the reason for these deferrals not happening (any ideas you have for that one would be great).

This is a tricky one! Any thoughts?

Reply:
It's possible to have missing names as MDEFs, and they should be skipped, so I'm now convinced that this is not the right fix. A more promising fix would be to separate `missing_names` per scope, since sharing it across scopes seems to be the underlying issue -- an incomplete definition in the module namespace blocks the addition of a thing to a class namespace. We have a TypedDict `T` and a nested `b.T`, and the common short name `T` is the key here. 

We want to be able to define `b.T` while the module-level `T` is incomplete, and we could do this if we'd recognize that `T` is not actually missing in the class scope/namespace (but it might be, so we can't just remove the check). This was a real head-scratcher!

Does this make sense to you? Would you like to try splitting `missing_names` per scope?

Reply:
Yeah that would make a lot more sense, I've had a go at implementing it but it wasn't super obvious. I wasn't totally sure on the best way to get the current scope to split on so any suggestions you have there would be interesting. 

I also looked into adding further tests to catch cases where this succeeds where the previous one failed. I thought a nested TypedDicts class could be a good way forward, but found that nested TypedDicts weren't being marked as incomplete. Any ideas you have for further testing would be great.

Reply:
Ping @JukkaL, any thoughts?

Reply:
I left a comment below. Your new approach works, but I have suggestion about how to make it more consistent with how other similar things are handled. And sorry for the slow response!

Reply:
No worries, thanks for the review!